### PR TITLE
base64 encoding/decoding module

### PR DESCRIPTION
**Pull request by BytesGalore**
This module provides encoding and decoding of base64 strings.

### CODE REVIEW

File: sys/base64/base64.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        else {
+            nNum += nLst << (8 - 2 * njump);
+            nLst =  tmpval & ((1 << njump * 2) - 1);
+        }
+
+        base64_out[iterate_base64_buffer++] = getsymbol(nNum);
+    }
+
+    // The last character is not finished yet
```
Comment (PeterKietzmann) [inline sys/base64/base64.c]: comment style

File: sys/base64/base64.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    njump++;
+
+    if (njump == 4) {
+        nNum = (tmpval >> (2 * njump));
+    }
+
+    nNum = nLst << (8 - 2 * njump);
+    base64_out[iterate_base64_buffer++] = getsymbol(nNum);
+
+    // if required we append '=' for the required dividability
```
Comment (PeterKietzmann) [inline sys/base64/base64.c]: comment style

File: sys/base64/base64.c [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    if (symbol == '/') {
+        return BASE64_SLASH;
+    }
+
+    if (symbol == '+') {
+        return BASE64_PLUS;
+    }
+
+    if (symbol == '=') {
+        // indicates a padded base64 end
```
Comment (PeterKietzmann) [inline sys/base64/base64.c]: comment style

Comment (PeterKietzmann) [inline sys/base64/base64.c]: and following...

**Comment by PeterKietzmann** [comment]
Besides some comment style issues the code looks good to me and the test works out of the box. For the rest I'm not deep enough into base64 but following the discussions all comments semm to be addressed. Someone willing to merge?

**Comment by BytesGalore** [comment]
@PeterKietzmann I've adjusted the comments

File: sys/base64/base64.c [hunk 4]
```diff
@@ -0,0 +1,227 @@
+/**
```
Comment (OlegHahm) [inline sys/base64/base64.c]: Still open (though not critical in a .c file).

File: sys/include/base64.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
+/**
```
Comment (OlegHahm) [inline sys/include/base64.h]: You might want do add an `@author` tag, too.

File: sys/include/base64.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_base64 base64 encoder decoder
+ * @ingroup     sys
+ * @brief       base64 encoder and decoder
+ * @{
+ *
+ * @file        base64.h
```
Comment (OlegHahm) [inline sys/include/base64.h]: omit the parameter for `@file`.

File: sys/include/base64.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#define BASE64_ERROR_DATA_IN_SIZE       (-4)    /**< error value for invalid output buffer size     */
+
+/**
+ * @brief           Encodes a given datum to base64 and save the result to the given destination.
+ * @param[in]       data_in           pointer to the datum to encode
+ * @param[in]       data_in_size      the size of `data_in`
+ * @param[out]      base64_out        pointer to store the encoded base64 string
+ * @param[in,out]   base64_out_size   pointer to the variable containing the size of `base64_out.`
+                                      This value is overwritten with the estimated size used for the encoded base64 string on BASE64_ERROR_BUFFER_OUT_SIZE.
+                                      This value is overwritten with the actual used size for the encoded base64 string on BASE64_SUCCESS.
```
Comment (OlegHahm) [inline sys/include/base64.h]: lines too long

File: tests/base64/main.c [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+#include "base64.h"
+
+int main(void)
+{
+    puts("-=== base64 tests ===-\n");
+
+    size_t data_in_size = 63;
+    unsigned char data_in[] = "Hello RIOT this is a base64 test!\nThis should work as intended.";
+
+    unsigned char expected_encoding[] = "SGVsbG8gUklPVCB0aGlzIGlzIGEgYmFzZTY0IHRlc3QhClRoaXMgc2hvdWxkIHdvcmsgYXMgaW50ZW5kZWQu";
```
Comment (OlegHahm) [inline tests/base64/main.c]: line is too long (suppressed for the the rest of this PR)

**Comment by OlegHahm** [comment]
I thinbk the test could be transformed into a unittest.

**Comment by BytesGalore** [comment]
@OlegHahm I've adjusted the linewidths and transformed the tests to unittest.

Seems that Travis has some problems here but I can't restart the build, though Im loged in.  
Anyone a hint for me?

File: tests/unittests/tests-base64/tests-base64.c [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+    printf("\nFrom:\n%s\n", (char *)encoded);
+#endif
+}
+
+Test *tests_base64_tests(void)
+{
+    EMB_UNIT_TESTFIXTURES(fixtures) {
+        new_TestFixture(test_base64_01_encode_string),
```
Comment (OlegHahm) [inline tests/unittests/tests-base64/tests-base64.c]: indentation

**Comment by OlegHahm** [comment]
Travis fail was unrelated. I don't know why you couldn't have restarted it yourself. Now, I did.

**Comment by BytesGalore** [comment]
@OlegHahm I fixed the indention (astyle tricked me).  
I still can't re-kick travis, the round-arrowed button left of the download-log button just doesn't show up.
Its like I'm not logged in, but I am (I logged out and in multiple times, but with no effect).

**Comment by PeterKietzmann** [comment]
I restarted Travis. Except for the need of squashing the build has been successful

**Comment by BytesGalore** [comment]
Hmm, funny I can't even edit/change the labels

**Comment by OlegHahm** [comment]
@BytesGalore, sorry, I think it was my fault. Some days ago I cleaned up the Github collaborator groups and deleted the "libcppa task force" group because I thought it's not needed any more. Apparently, your account was only assigned to this group, but not to the regular maintainer group (which I fixed just now). Sorry again!

**Comment by BytesGalore** [comment]
no problem at all, thx for adding me :smile: .

**Comment by OlegHahm** [comment]
ACK and go.

**Comment by OlegHahm** [comment]
(Congrats to commit number 5000! :cake:

**Comment by BytesGalore** [comment]
nice :tada:

**Pull request merged**